Create attribute within the new product MVP

louwie17 opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

This allows users to create a new attribute if the name they are searching for doesn't exist within the attribute list.
This is currently dependent on #34999 to be merged, which is why it's pointing at that branch.

https://user-images.githubusercontent.com/2240960/195870763-a609958f-b3ab-4515-a498-25660423b5c9.mp4

Address the creation part in #34331  .

### How to test the changes in this Pull Request:

1. Load this branch, build it, and make sure you have the `new-product-management-experience` feature flag enabled (you can do so using the latest version of the Beta tester within the mono repo).
2. Go to **Products > Add new (MVP)** and fill out some of the initial details like name and price
3. Scroll down and click **Add first attribute** in the attributes section
4. Now write a random name in the attribute search field in the table, of an attribute that doesn't exist.
5. The 'Create "your new attribute name"' should show as an option
6. Click that and the new item should show as selected and the term field should focus
7. With `localStorage.setItem( 'debug', 'wc-admin:*' );` run prior, it should show that a `product_attribute_add_custom_attribute` has been fired.
8. Start adding attribute terms, this dropdown should initial be blank and any text you put in will show up as a `Create "text"` option, allowing you to add several new items.
9. Click **Add** the custom attribute should show correctly in the main list, now hit `Update` or `Publish`
10. Go to old product editor and load the same product you just saved, check the attributes tab and make sure the custom one's show up correctly.

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [x] Have you written new tests for your changes, as applicable?
-   [x] Have you successfully run tests with your changes locally?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> run changelog add`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

louwie17 committed:
Add initial add new option

louwie17 committed:
Hook in create attribute modal to add attribute field

louwie17 committed:
Add unit tests for the create attribute modal

louwie17 committed:
Add extra test to attribute input field

louwie17 commented on plugins/woocommerce-admin/client/products/fields/attribute-input-field/attribute-input-field.tsx:
```diff
@@ -65,51 +88,80 @@ export const AttributeInputField: React.FC< AttributeInputFieldProps > = ( {
 		: null;
 
 	return (
-		<SelectControl< Pick< QueryProductAttribute, 'id' | 'name' > >
```
This is just indented as I had to add a Fragment around this in order to add the create modal

louwie17 committed:
Add changelog

louwie17 committed:
Add custom attribute term input field and support for custom attributes

louwie17 committed:
Fix tets

louwie17 committed:
Add track for custom attribute creation

louwie17 committed:
Fix changes after merge conflict

louwie17 committed:
Revert one change

louwie17 committed:
Fix lint error

joshuatf commented on plugins/woocommerce-admin/client/products/fields/attribute-field/add-attribute-modal.tsx:
```diff
@@ -32,9 +35,15 @@ type AddAttributeModalProps = {
 };
 
+function isEmptyItem(
+	item: HydratedAttributeType | { id: undefined }
```
Could this be a partial of `HydratedAttributeType` or does it need to have a separate `{ id: undefined }`?

louwie17 replied:
I wanted to have a clear differentiator between when a attribute was set and when it wasn't. Using a Partial of HydratedAttributeType would make that a bit more difficult and we did have to manually check if all the attributes exist on that property to deem it as the HydratedAttributeType in several places.

louwie17 replied:
I could potentially just try `null`, let me see if that works well.

louwie17 replied:
I updated it to using `HydratedAttributeType | null` so that things are more clear.

joshuatf replied:
Awesome!

joshuatf commented on plugins/woocommerce-admin/client/products/fields/attribute-field/add-attribute-modal.tsx:
```diff
@@ -186,12 +210,15 @@ export const AddAttributeModal: React.FC< AddAttributeModalProps > = ( {
-																		options:
-																			undefined,
-																	}
+																	typeof val ===
```
It would be great to have this as a util method so we could also test it.

E.g., `parseAttribute( val: string | Attribute )`

louwie17 replied:
Good suggestion, I did do this as part of the https://github.com/woocommerce/woocommerce/pull/35100/commits/fd8af0d7fe3a812dbff3f5ace8077e099f6298f8 commit (it includes the other changes as well)

joshuatf commented on plugins/woocommerce-admin/client/products/fields/attribute-input-field/attribute-input-field.tsx:
```diff
@@ -76,11 +104,20 @@ export const AttributeInputField: React.FC< AttributeInputFieldProps > = ( {
-					name: attribute.name,
-					options: [],
-				} );
+				if ( attribute.id === -99 ) {
```
To reduce the number of `attribute.id === -99` checks, maybe we could create a util method `isNewAttribute( attribute )` that handles this check for us.

louwie17 replied:
Done as part of: https://github.com/woocommerce/woocommerce/pull/35100/commits/2dcbcd50c61e66e3a28be45f78d9fc2101674947

joshuatf commented on plugins/woocommerce-admin/client/products/fields/attribute-field/add-attribute-modal.tsx:
```diff
@@ -219,31 +246,63 @@ export const AddAttributeModal: React.FC< AddAttributeModalProps > = ( {
-																		index +
-																		'].terms',
+														{ isEmptyItem(
+															formAttr
```
I know that `formAttr` was not added in this PR, but this confused me for a minute as I thought it was a "form attribute."   Would `attribute` be a more accurate name?  Or is there something special about these that is related to the form?

joshuatf reviewed (dismissed):
Very nice work, Lourens!  Testing well on my end; left a few minor comments, but nothing blocking and pre-approving this PR.

Not a result of this PR, but I noticed a couple of oddities around styling:

1. The modal appears off when editing an existing attribute
<img width="642" alt="Screen Shot 2022-11-16 at 7 44 44 AM" src="https://user-images.githubusercontent.com/10561050/202227761-13f01c6f-0e9c-45ec-9a5c-74de763043ae.png">

2.  The loading indicator is slightly disorienting and jumpy, making it challenging to click an item while loading occurs.  This loading indicator might be better as the select control's icon instead of in the list (cc @jarekmorawski).  Though this may already be planned and blocked by the select control icon issue.
<img width="243" alt="Screen Shot 2022-11-16 at 7 43 45 AM" src="https://user-images.githubusercontent.com/10561050/202227898-ec3309d8-f032-4cb5-b931-5f93b9d99b3e.png">

jarekmorawski commented:
> The loading indicator is slightly disorienting and jumpy, making it challenging to click an item while loading occurs. This loading indicator might be better as the select control's icon instead of in the list (cc @jarekmorawski). Though this may already be planned and blocked by the select control icon issue.

Definitely. The latest approach is to replace the search icon with a loading indicator, but we'll work on that as part of another issue.

louwie17 committed:
Seperate out some logic and make use of null as empty object versus id: undefined

louwie17 committed:
Add isNewAttributeListItem helper function

louwie17 commented:
@joshuatf this should be ready for re-review, I addressed all your changes.
Also in relation to the two issues you found:

> The modal appears off when editing an existing attribute

This should be fixed now after the rebase as I fixed this as part of the create attribute PR -> https://github.com/woocommerce/woocommerce/pull/35132 (at-least I can't reproduce it anymore)

> The loading indicator is slightly disorienting and jumpy, making it challenging to click an item while loading occurs. This loading indicator might be better as the select control's icon instead of in the list (cc @jarekmorawski). Though this may already be planned and blocked by the select control icon issue.

As @jarekmorawski mentioned this is blocked by the other issue in flight to allow us to move the loading indicator within the input field.

joshuatf commented on plugins/woocommerce-admin/client/products/fields/attribute-input-field/attribute-input-field.tsx:
```diff
@@ -104,7 +145,27 @@ export const AttributeInputField: React.FC< AttributeInputFieldProps > = ( {
 									getItemProps={ getItemProps }
 								>
-									{ item.name }
+									{ item.id === -99 ? (
```
This could also use the new `isNewAttributeListItem` right?

louwie17 replied:
Nice catch, not sure how I missed that, fixed here: https://github.com/woocommerce/woocommerce/pull/35100/commits/9a7f07b479e490434f630bb78a541959d40f7874

joshuatf commented on plugins/woocommerce-admin/client/products/fields/attribute-field/add-attribute-modal.tsx:
```diff
@@ -83,28 +91,28 @@ export const AddAttributeModal: React.FC< AddAttributeModalProps > = ( {
-			setTimeout( () => {
+		setTimeout( () => {
+			const valueInputField: HTMLInputElement | null =
+				document.querySelector(
```
I think ideally we could use a ref here instead of dom selectors, but we can save that for a follow-up issue as I know that wasn't really a part of this PR.

louwie17 replied:
Yeah currently the SelectControl component doesn't allow us to pass in a `ref`, I could definitely add that though as a separate PR

joshuatf reviewed (dismissed):
Thanks for the changes, @louwie17!  This is still testing well for me and changes look great.

Left one more comment about using refs instead of dom selectors for a possible follow-up issue.  Otherwise, LGTM!  🚢

louwie17 committed:
Make use of helper function for create new markup

louwie17 commented:
@joshuatf could I get a re-review/approval I just updated the one last spot that needed to use the helper function.

joshuatf reviewed (approved):
Changes look good!  LGTM :shipit:
